### PR TITLE
Refactor/algorithms in stac api

### PR DESCRIPTION
**Pull request by Thuhaa**
Thank you for submitting a pull request!

## Description

Implemented algorithms setting in StacApiExplorer component

image

image

### Type of Pull Request

- [ ] Adding a feature
- [ ] Fixing a bug
- [ ] Maintaining documents
- [ ] Adding tests
- [ ] Others ()

### Verify the followings

- [ ] Code is up-to-date with the `develop` branch
- [ ] No build errors after `pnpm build`
- [ ] No lint errors after `pnpm lint`
- [ ] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`
- [ ] Make sure all the existing features working well

### Changesets

- [ ] If your PR makes a change under `packages` folder that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

Refer to [CONTRIBUTING.MD](https://github.com/UNDP-Data/geohub/blob/develop/CONTRIBUTING.md) for more information.

### CODE REVIEW

File: sites/geohub/src/lib/server/helpers/createDatasetLinks.ts [hunk 1]
```diff
@@ -211,41 +213,81 @@ export const createDatasetLinks = async (
...
+					type: 'application/json',
+					href: `${titilerUrl}/WebMercatorQuad/tilejson.json?url=${encodeURIComponent(
+						b64EncodedUrl
+					)}&scale=1&resampling=nearest&return_mask=true${algorithmId ? '' : '&bidx=1'}&algorithm=${tool}`
+				});
+				feature.properties.links.push({
+					rel: 'algorithms',
+					type: 'application/json',
+					href: `${titilerUrl.replace('cog', 'algorithms')}`
+				});
```
Comment (JinIgarashi) [inline sites/geohub/src/lib/server/helpers/createDatasetLinks.ts:254]: Can you remove these duplicated codes? It looks same with line 255-290 apart from info and statistics endpoint.

You can merge both of them. also `info` endpoint does not have algorithm query param.

File: sites/geohub/src/routes/api/stac/[id]/[collection]/[...item]/tools/[tool]/+server.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import type { RequestHandler } from '@sveltejs/kit';
+import { createDatasetLinks } from '$lib/server/helpers';
+import { env } from '$env/dynamic/private';
+
+export const POST: RequestHandler = async ({ request, url, params }) => {
+	const feature = await request.json();
+
+	const tool = params.tool;
+
+	feature.properties = await createDatasetLinks(feature, url.origin, env.TITILER_ENDPOINT, tool);
```
Comment (JinIgarashi) [inline sites/geohub/src/routes/api/stac/[id]/[collection]/[...item]/tools/[tool]/+server.ts:10]: do we need this endpoint?

I think `algorithmId` can be included inside feature object as tag. you can get algorithmId from tags.

**Review by JinIgarashi — CHANGES_REQUESTED** [review]
@Thuhaa So far, I commented two things. could you fix them? also remove also comments and console.log before you merge it to develop.

File: sites/geohub/src/components/util/stac/StacApiExplorer.svelte [hunk 3]
```diff
@@ -291,6 +311,15 @@
...
 					stacDatasetFeature = await getProductFeature(itemIds);
 				}
+				if (selectedTool) {
+					for (const f of clickedFeatures.slice(0, clickedFeatures.length - 1)) {
+						map.setFeatureState(f, { click: false });
+					}
+					clickedFeatures = [clickedFeatures.at(-1)];
+					const itemIds = clickedFeatures.map((f) => f.properties.id);
+					map.setFeatureState(clickedFeatures[0], { click: true });
+					stacDatasetFeature = await getToolsFeature(itemIds);
```
Comment (Thuhaa) [inline sites/geohub/src/components/util/stac/StacApiExplorer.svelte:321]: @JinIgarashi  The stacDatasetFeature returned here doesn't seem to work with the MiniMap component to show a preview

Comment (JinIgarashi) [reply sites/geohub/src/components/util/stac/StacApiExplorer.svelte:321]: @Thuhaa If it is not work, could you fix it?

**Comment by JinIgarashi** [comment]
@Thuhaa 

> The showing of the preview in the minimap fails when an item is selected

Could you describe more precisely? How does it not work? What did you try to debug?

File: .changeset/dirty-seahorses-look.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+"geohub": minor
+---
+
+feat: registering of algorithms
```
Comment (JinIgarashi) [inline .changeset/dirty-seahorses-look.md:5]: please merge two changeset files into a file. a PR should only have a changeset file for each package/site.

File: sites/geohub/src/components/util/MiniMap.svelte [hunk 5]
```diff
@@ -24,6 +24,9 @@
 	export let isLoadMap = false;
 	export let layer: VectorLayerTileStatLayer = undefined;
 	export let band: string = undefined;
+	export let tool: string = undefined;
```
Comment (JinIgarashi) [inline sites/geohub/src/components/util/MiniMap.svelte:27]: I think you decided not to change MiniMap (preview) for tools now. Are these changes necessary for this PR? could you revert them to original one?

File: sites/geohub/src/components/util/stac/StacCatalogTool.svelte [hunk 6]
```diff
@@ -97,7 +97,6 @@
 					link.href = newLink.href;
 				}
 			});
-
 			const assetTags = Object.keys(selectedAssets).map((name) => {
```
Comment (JinIgarashi) [inline sites/geohub/src/components/util/stac/StacCatalogTool.svelte:100]: add an empty row here. try to minimise file change

File: sites/geohub/src/lib/RasterTileData.ts [hunk 7]
```diff
@@ -145,10 +146,12 @@ export class RasterTileData {
 							.join('&')}`
 					: ''
 			}`;
+
 			const res = await fetch(apiUrl, {
 				method: 'POST',
 				body: data
 			});
+
 			savedLayerStyle = await res.json();
```
Comment (JinIgarashi) [inline sites/geohub/src/lib/RasterTileData.ts:155]: remove these three lines empty row from this file. try to minimise file change

File: sites/geohub/src/lib/server/defaultStyle/RasterDefaultStyle.ts [hunk 8]
```diff
@@ -204,8 +204,11 @@ export default class RasterDefaultStyle implements DefaultStyleTemplate {
 	};
 
 	public getMetadata = async (algorithmId?: string) => {
+		// console.log(this.dataset.properties?.links)
 		const metadataUrl = this.dataset.properties?.links?.find((l) => l.rel === 'info').href;
 		const product = this.dataset.properties.tags?.find((t) => t.key === 'product')?.value;
+		// console.log("ZXXXXXXXXXXXXXXXX", metadataUrl)
+		// console.log("DDDDDDDDDDDDDDDDD", this.metadata)
```
Comment (JinIgarashi) [inline sites/geohub/src/lib/server/defaultStyle/RasterDefaultStyle.ts:211]: remove comments, please

File: sites/geohub/src/lib/server/helpers/createDatasetLinks.ts [hunk 9]
```diff
@@ -5,13 +5,16 @@ import { generateAzureBlobSasToken } from './generateAzureBlobSasToken';
 export const createDatasetLinks = async (
 	feature: DatasetFeature,
 	origin: string,
-	titilerUrl: string
+	titilerUrl: string,
+	tool?: string
```
Comment (JinIgarashi) [inline sites/geohub/src/lib/server/helpers/createDatasetLinks.ts:9]: Is this `tool` param necessary in this function? `tool (algorithm ID)` can be fetched from tags like what you are doing in line 14-17

File: sites/geohub/src/routes/(app)/management/stac/api/[id]/+layout.server.ts [hunk 10]
```diff
@@ -6,17 +6,28 @@ import { generateHashKey } from '$lib/helper';
...
+
+	const isRegistered = datasetRes.status !== 404;
+
 	if (!stac) {
 		error(404, `This stac ID (${id}) is not found.`);
 	}
+
 	return {
 		stac,
+		dataset,
```
Comment (JinIgarashi) [inline sites/geohub/src/routes/(app)/management/stac/api/[id]/+layout.server.ts:30]: I cannot understand why you need to return dataset  at +layout.server.ts in stac management pages.

if dataset is necessary for `tool`, you can fetch dataset inside StacApiExplorer component when users select Tools tab. I think that is waht current STAC api component is doing for assets and product.

So, you basically don't need to export  `dataset` prop from StacApiExplorer

File: sites/geohub/src/routes/(app)/management/stac/api/[id]/[collection]/+page.svelte [hunk 11]
```diff
@@ -132,6 +133,7 @@
 		<p class="title is-5">STAC data explorer</p>
 
 		<StacApiExplorer
+			bind:dataset
```
Comment (JinIgarashi) [inline sites/geohub/src/routes/(app)/management/stac/api/[id]/[collection]/+page.svelte:136]: try remove this.

File: sites/geohub/src/routes/api/stac/[id]/[collection]/[...item]/tools/[tool]/+server.ts [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+export const POST: RequestHandler = async ({ request, url }) => {
+	const feature = await request.json();
+
+	// const tool = params.tool;
+
+	feature.properties = await createDatasetLinks(feature, url.origin, env.TITILER_ENDPOINT);
+
+	console.log(feature);
+	return new Response(JSON.stringify(feature));
+};
```
Comment (JinIgarashi) [inline sites/geohub/src/routes/api/stac/[id]/[collection]/[...item]/tools/[tool]/+server.ts:14]: looks like this API is not used. remove this.

**Review by JinIgarashi — CHANGES_REQUESTED** [review]
@Thuhaa I commented. please fix them. try to minimise file changes.

File: sites/geohub/src/lib/server/helpers/createDatasetLinks.ts (hunk 9 above)
Comment (Thuhaa) [reply sites/geohub/src/lib/server/helpers/createDatasetLinks.ts:9]: @JinIgarashi Some datasets might hold more than one tool. So in this case, would it be advisable to have tool as a tag? Fetching from tags might not work properly then I believe

File: sites/geohub/src/routes/api/stac/[id]/[collection]/[...item]/tools/[tool]/+server.ts (hunk 12 above)
Comment (Thuhaa) [reply sites/geohub/src/routes/api/stac/[id]/[collection]/[...item]/tools/[tool]/+server.ts:14]: I have used it in the `getToolsFeature` function of `StacApiExplorer` component

File: sites/geohub/src/components/util/MiniMap.svelte [hunk 13]
```diff
@@ -110,7 +112,6 @@
 					}
 					const data = await rasterTile.add(map, bandIndex);
 					metadata = data.metadata;
-
 					dispatch('layerAdded', data);
```
Comment (JinIgarashi) [inline sites/geohub/src/components/util/MiniMap.svelte:115]: could you revert these changes for MiniMap?

File: sites/geohub/src/routes/(app)/data/[id]/+page.svelte [hunk 14]
```diff
@@ -329,14 +329,20 @@
 							{@const isCatalog =
 								feature.properties.tags.find((t) => t.key === 'stacApiType')?.value === 'catalog'}
 
+							<!-- -->
 							{#if isCatalog}
```
Comment (JinIgarashi) [inline sites/geohub/src/routes/(app)/data/[id]/+page.svelte:333]: delete this comment

**Review by JinIgarashi — COMMENTED** [review]
can you revert 2 files changes I commented?

File: sites/geohub/src/routes/api/stac/[id]/[collection]/[...item]/tools/[tool]/+server.ts [hunk 15]
```diff
@@ -0,0 +1,9 @@
+import type { RequestHandler } from '@sveltejs/kit';
+import { createDatasetLinks } from '$lib/server/helpers';
+import { env } from '$env/dynamic/private';
+
+export const POST: RequestHandler = async ({ request, url }) => {
+	const feature = await request.json();
+	feature.properties = await createDatasetLinks(feature, url.origin, env.TITILER_ENDPOINT);
+	return new Response(JSON.stringify(feature));
```
Comment (JinIgarashi) [inline sites/geohub/src/routes/api/stac/[id]/[collection]/[...item]/tools/[tool]/+server.ts:8]:
```suggestion
	const feature = await request.json();
	const tool = params.tool
	feature.properties.tags = feature.propterites.tags.filter(t=>t.key!=='algorithm')
	feature.properties.tags = [...feature.properties.tags, {"key": "algorithm", "value": tool}]
	feature.properties = await createDatasetLinks(feature, url.origin, env.TITILER_ENDPOINT);
	return new Response(JSON.stringify(feature));
```

File: sites/geohub/src/routes/api/stac/[id]/[collection]/[...item]/tools/[tool]/+server.ts [hunk 16]
```diff
@@ -0,0 +1,9 @@
+import type { RequestHandler } from '@sveltejs/kit';
```
Comment (JinIgarashi) [inline sites/geohub/src/routes/api/stac/[id]/[collection]/[...item]/tools/[tool]/+server.ts:1]: rename `tool` to `algorithm`

File: sites/geohub/src/components/util/stac/StacApiExplorer.svelte [hunk 17]
```diff
@@ -538,9 +590,85 @@
...
+			}
+		});
+
+		feature.properties.tags;
+		feature.properties.tags = [
+			{ key: 'type', value: 'api' },
+			...feature.properties.tags,
+			...assetTags
+		];
+		const url = `/api/stac/${stacId}/${collection}/${ids.join('/')}/tools/${selectedAlgorithmName}`;
```
Comment (JinIgarashi) [inline sites/geohub/src/components/util/stac/StacApiExplorer.svelte:649]: rename tools to algorithms

**Review by JinIgarashi — APPROVED** [review]
